File: js/test/runtimecontrol_example.ts
import * as RTI from "../src/index.js"

let simTime = 0.0

function update(dt: number) {
    console.log(`Update: simTime=${simTime.toFixed(3)} dt=${dt.toFixed(3)}`)
    // do simulation work
}

const rti = new RTI.Client({ application: "typescript_runtimecontrol_example" })
const runtime = new RTI.RuntimeControl(rti)

runtime.onStart = () => { console.log("Simulation started") }
runtime.onStop = () => { console.log("Simulation stopped") }
runtime.onEnd = () => { console.log("Simulation ended") }
runtime.onReset = () => { console.log("Simulation reset"); simTime = 0.0 }
runtime.onLoadScenario = (scenario, playback) => {
    console.log(`Loading scenario: ${scenario.name}`)
    return true
}
runtime.onTimeScale = (ts) => { console.log(`Time scale: ${ts}`) }

rti.on("connect", () => { console.log("Connected") })
rti.on("disconnect", () => { console.log("Disconnected") })
rti.on("error", (e: any) => { console.error("Error:", e) })

// Simulation loop using setInterval (Node.js real-time pattern)
let lastTime = Date.now()
const interval = setInterval(() => {
    if (rti.state === RTI.proto.RuntimeState.RUNNING) {
        const now = Date.now()
        const dt = (now - lastTime) / 1000
        lastTime = now
        simTime += dt * (runtime.timeScale ?? 1.0)
        update(dt)
    } else {
        lastTime = Date.now() // keep lastTime current when idle
    }
}, 10)

process.on("SIGINT", () => {
    clearInterval(interval)
    rti.disconnect()
    setTimeout(() => process.exit(0), 200)
})

File: python/test/runtimecontrol_example.py
import sys, os
sys.path.insert(0, os.path.dirname(__file__) + "/..")

import inhumate_rti as RTI
import time

sim_time = 0.0

def update(dt):
    print(f"Update: {sim_time} {dt}")
    # do simulation work
    time.sleep(0.1)

last_real_time = time.time()

def main_loop():
    global sim_time
    global last_real_time

    if rti.state == RTI.proto.RuntimeState.RUNNING:
        real_time = time.time()
        dt = real_time - last_real_time
        last_real_time = real_time
        sim_time += dt * (runtime.time_scale or 1.0)
        update(dt)
    else:
        time.sleep(0.1) # idle

rti = RTI.Client(application="python_usage_example", main_loop=main_loop)
runtime = RTI.RuntimeControl(rti)

def on_text(text: str):
    print(f"Received text: {text}")
rti.subscribe_text("text", on_text)

def on_connect():
    print("Connected")
rti.on("connect", on_connect)

def on_disconnect():
    print("Disconnected")
rti.on("disconnect", on_disconnect)

def on_error(type, message, exception):
    print(f"Error: {type}: {message} {exception}")
    # traceback.print_stack()
rti.on("error", on_error)


try:
    rti.connect() # now blocking with main loop

except KeyboardInterrupt:
    pass
rti.disconnect()
